doc/developer/design/20240610_unified_compute_introspection.md

```diff
@@ -0,0 +1,134 @@
+We propose to give the compute controller access to the coordinator's `transient_id_gen` by converting it from an `u64` to an `Arc<AtomicU64>` and passing a reference to the compute controller during creation.
+
+The compute controller needs to handle responses for introspection subscribes specially:
+Rather than propagating them to its clients, it needs to translate them into updates of the associated storage-managed collections:
+
+- `Row` updates (both insertions and retractions) are translated into the schema of the storage-managed collection by prepending the replica ID as the first column.
+  The modified `Row`s are then sent to the storage controller for recording under the respective `IntrospectionType`.
+- Subscribe errors and cancellations are translated into deletion requests to the storage controller.
+  A deletion request targets the storage-managed collection identified by the respective `IntrospectionType` and contains a filter on the first column of each contained `Row` that selects the ID of the replica whose introspection subscribe was interrupted.
+
+### Deletion from Storage-managed Collections
+
+The storage controller delegates the writing of storage-managed collections to a background task called the `CollectionManager`.
+The `CollectionManager` accepts `GlobalId`s identifying storage collections and corresponding `(Row, Diff)` updates to be appended to these collections, and appends them at the current system time.
+It does not yet expose a mechanism to delete previously written updates from a collection by reading back the collection contents, determining the necessary retractions, and appending them.
+
+To remove introspection data for dropped/disconnected replicas from the unified introspection relations, we require such a deletion mechanism.
+Note that the compute controller is not itself able to determine the set of necessary retractions.
+While it can be taught to read the contents of storage-managed collections, it doesn't have any way of knowing the time as of which all previously emitted introspection data has been fully written to its target collection.
+The compute controller would risk reading back the target collection's contents too soon and retract only part of the introspection data previously emitted for the disconnected replica.
+
+We propose extending the `CollectionManager`'s API to also accept deletion requests.
+A deletion request contains a collection ID, as well as a `filter` closure `Box<dyn Fn(&Row) -> bool>`.
+The `CollectionManager` handles a deletion request by reading back the target collection at its latest readable time, applying the `filter` closure to each read `Row`, and appending retractions for each `Row` the `filter` returned `true` for.
+Append and deletion requests must be applied by the `CollectionManager` in the same order they have been issued by the client, to guarantee that deletions will retract all previously sent appends.
```

Comment:
What's the memory requirements for providing such a feature?

Reply:
This depends a lot on the implementation of the persist client.

With the naive implementation (build the retractions in memory, then append them):
* In the "best case" the persist client streams in one update at a time and we only need to keep those belonging to the deleted replica, so the memory usage would be equal to the size of existing updates of that replica.
* In the realistic case, the persist client probably introduces some overhead because it gives us the updates in batches, not one at a time. We can hope that this is constant overhead, but I'm not sure.
* In the worst case we'd need to keep the whole collection snapshot in memory.

Rather than collecting retractions in memory, we can use a `BatchBuilder`, which will flush out parts to S3 when they get too large. If we make sure the unified introspection collection stay small that's probably not an optimization we need.

Also note that the current Pv2 plan includes making the `CollectionManager` self-correcting (https://github.com/MaterializeInc/materialize/pull/27496), similar to the persist_sink, in which case it will need to read the whole snapshot into memory anyway.